### ISSUE TITLE
"Error launching CUDA compiler: 256" on OpenMM context creation

### ISSUE THREAD

**Issue by andrrizzi** [choderalab/yank#347]
Keep having this error on the cluster running with `nodes=2:ppn=4:gpus=4:shared`. The exception is raised after the solvent phase has been completed, during resuming of the complex phase. I'm looking into this.

```
Traceback (most recent call last):
  File "/cbio/jclab/home/andrrizzi/miniconda/bin/yank", line 9, in <module>
    load_entry_point('yank==0.9.0', 'console_scripts', 'yank')()
  File "/cbio/jclab/home/andrrizzi/miniconda/lib/python2.7/site-packages/yank-0.9.0-py2.7-linux-x86_64.egg/yank/cli.py", line 105, in main
    dispatched = getattr(commands, command).dispatch(args)
  File "/cbio/jclab/home/andrrizzi/miniconda/lib/python2.7/site-packages/yank-0.9.0-py2.7-linux-x86_64.egg/yank/commands/script.py", line 34, in dispatch
    yaml_builder.build_experiment()
  File "/cbio/jclab/home/andrrizzi/miniconda/lib/python2.7/site-packages/yank-0.9.0-py2.7-linux-x86_64.egg/yank/yamlbuild.py", line 1203, in build_experiment
    self._run_experiment(combination, output_dir)
  File "/cbio/jclab/home/andrrizzi/miniconda/lib/python2.7/site-packages/yank-0.9.0-py2.7-linux-x86_64.egg/yank/yamlbuild.py", line 1778, in _run_experiment
    yank.run()
  File "/cbio/jclab/home/andrrizzi/miniconda/lib/python2.7/site-packages/yank-0.9.0-py2.7-linux-x86_64.egg/yank/yank.py", line 451, in run
    simulation.run(niterations_to_run=niterations_to_run)
  File "/cbio/jclab/home/andrrizzi/miniconda/lib/python2.7/site-packages/yank-0.9.0-py2.7-linux-x86_64.egg/yank/repex.py", line 827, in run
    self._initialize_resume()
  File "/cbio/jclab/home/andrrizzi/miniconda/lib/python2.7/site-packages/yank-0.9.0-py2.7-linux-x86_64.egg/yank/repex.py", line 1011, in _initialize_resume
    self.platform = self._determine_fastest_platform(representative_system)
  File "/cbio/jclab/home/andrrizzi/miniconda/lib/python2.7/site-packages/yank-0.9.0-py2.7-linux-x86_64.egg/yank/repex.py", line 901, in _determine_fastest_platform
    context = openmm.Context(system, integrator)
  File "/cbio/jclab/home/andrrizzi/miniconda/lib/python2.7/site-packages/simtk/openmm/openmm.py", line 15103, in __init__
    this = _openmm.new_Context(*args)
Exception: Error launching CUDA compiler: 256
<built-in>:0:0: fatal error: when writing output to : Bad file descriptor
compilation terminated.
```

**Comment by jchodera** [comment]
One other thought: I think the new openmm-dev requires cuda 7.5. I may not
have updated my module environment yet.

**Comment by andrrizzi** [comment]
Thanks! I'll give it a try.

**Comment by andrrizzi** [comment]
Currently, conda installing ``yank-dev`` install ``openmm 6.3.1``, not ``openmm-dev``, so I guess cuda 7.0 should work in my environment. Unless ``yank-dev`` have problems with the last ``openmm`` release.

**Comment by jchodera** [comment]
Let me know if you can also dump the `$PBS_GPUFILE` for this.

**Comment by andrrizzi** [comment]
I did ``cat $PBS_GPUFILE`` at the beginning of the simulation, before running ``yank``. I got
```
gpu-2-8-gpu3
gpu-2-8-gpu2
gpu-2-8-gpu1
gpu-2-8-gpu0
gpu-1-13-gpu3
gpu-1-13-gpu2
gpu-1-13-gpu1
gpu-1-13-gpu0
```

**Comment by andrrizzi** [comment]
The error seems to have disappeared with no apparent changes. I'm confused.

At this point I think this is some node on the cluster that is not configured correctly, or a bug affecting a particular GPU card model. I run without problems on ``gpu-2-8`` and ``gpu-1-5``, but I got the CUDA compiler error on ``gpu-2-8`` and ``gpu-1-13``, so now I'm trying to run again on ``gpu-1-13`` to see if I can reproduce the error.

**Comment by jchodera** [comment]
This must be a misconfigured node...

**Comment by andrrizzi** [comment]
I'm guessing this was related to #367, closing for now.

**Comment by jchodera** [comment]
I'm still having this problem with CUDA on `hal`. Is anyone else seeing this?

**Comment by Lnaden** [comment]
What GPU is it running on? I didn't think we ever solved that problem for running CUDA 8.0 on the older cards with the OpenMM compiled off of it.

**Comment by Lnaden** [comment]
This issue [came up](https://github.com/pandegroup/openmm/issues/95) back in '13 apparently. Although I don't know if its the same underlying problem.

**Comment by Lnaden** [comment]
Can you point to what system, yank version, and harwdware, you are using so I can try and debug a bit more. I'm willing to bed it has something to do with the forces but I cannot replicate myself

**Comment by jchodera** [comment]
I was running abl-imatinib from yank-examples using CUDA, with latest development yank, on only GTX680s. All nodes had the same Nvidia driver and CUDA 8 installed.

Just wondering if any of you can get this to work on hal. If so, I can debug further.

**Comment by Lnaden** [comment]
I can run the abl-imatinib examples without crashing on the context creation step. I using the head version for YANK (e767419) and the openmm 7.2.0 beta which has the most recent OpenCL fix in.

**Comment by Lnaden** [comment]
Short version: The problem is a race condition with the CUDA platform that happens on multiple clusters, but it will be harder to fix than we thought.

Long version:
The error occurs on under the following conditions: MPI, CUDA platform, accessing CUDA based code. I was able to confirm this happens on Hal (PBS), Lila (LSF), and the new GPU cluster at Merck (SLURM). So it is not a scheduler problem.

At first we thought it was just context creation so I created a blocking MPI send/receive function for context [in a new branch](https://github.com/choderalab/yank/blob/a50f282c49f51a768275ec4ae08e3b10557a8023/Yank/repex.py#L1172-L1209) for testing. This stopped the race condition on Context creation, but then failed immediately after:

```shell
# Partial backtrace
  File "/home/nadenl/miniconda3/envs/yex/lib/python3.5/site-packages/yank/repex.py", line 313, in _compute_potential_energy
    potential_energy = context.getState(getEnergy=True).getPotentialEnergy()
  File "/home/nadenl/miniconda3/envs/yex/lib/python3.5/site-packages/simtk/openmm/openmm.py", line 3969, in getState
    self._getStateAsLists(getP, getV, getF, getE, getPa, getPd, enforcePeriodic, groups_mask)
  File "/home/nadenl/miniconda3/envs/yex/lib/python3.5/site-packages/simtk/openmm/openmm.py", line 3919, in _getStateAsLists
    return _openmm.Context__getStateAsLists(self, getPositions, getVelocities, getForces, getEnergy, getParameters, getParameterDerivatives, enforcePeriodic, groups)
Exception: Error launching CUDA compiler: 256
<built-in>:0:0: fatal error: when writing output to : Bad file descriptor
```

Based on this output, I think we would have to wrap every single CUDA GPU operation in the MPI blocking send/receive to allow normal execution on the CUDA platform. Which pretty much defeats the purpose of using MPI at all. I think before we elevate this to the main OpenMM GitHub, there are 2 things that need to happen:
1. Simple test case which does context creation, energy computation, 1 step of propagation and energy computation (easy enough)
2. Non-cluster based local machine with CUDA platform that we have full write access on. This way we can test if its a permissions problem or a CUDA + MPI + OpenMM problem.


Full backtrace + lines which used to cause failure

```shell
2017-02-23 09:49:08,320: Creating and caching Context and Integrator. # <---- Previous failure point
2017-02-23 09:49:09,043: Context creation took 0.723 s.
Traceback (most recent call last):
  File "/home/nadenl/miniconda3/envs/yex/bin/yank", line 11, in <module>
    load_entry_point('yank==0.15.1', 'console_scripts', 'yank')()
  File "/home/nadenl/miniconda3/envs/yex/lib/python3.5/site-packages/yank/cli.py", line 71, in main
    dispatched = getattr(commands, command).dispatch(command_args)
  File "/home/nadenl/miniconda3/envs/yex/lib/python3.5/site-packages/yank/commands/script.py", line 100, in dispatch
    yaml_builder.build_experiments()
  File "/home/nadenl/miniconda3/envs/yex/lib/python3.5/site-packages/yank/yamlbuild.py", line 1333, in build_experiments
    self._run_experiment(combination, output_dir)
  File "/home/nadenl/miniconda3/envs/yex/lib/python3.5/site-packages/yank/yamlbuild.py", line 2528, in _run_experiment
    yank.run()
  File "/home/nadenl/miniconda3/envs/yex/lib/python3.5/site-packages/yank/yank.py", line 584, in run
    simulation.run(niterations_to_run=niterations_to_run)
  File "/home/nadenl/miniconda3/envs/yex/lib/python3.5/site-packages/yank/repex.py", line 852, in run
    self._initialize_resume()
  File "/home/nadenl/miniconda3/envs/yex/lib/python3.5/site-packages/yank/repex.py", line 1070, in _initialize_resume
    self._minimize_and_equilibrate()
  File "/home/nadenl/miniconda3/envs/yex/lib/python3.5/site-packages/yank/repex.py", line 1428, in _minimize_and_equilibrate
    self._minimize_replica(replica_index)
  File "/home/nadenl/miniconda3/envs/yex/lib/python3.5/site-packages/yank/sampling.py", line 619, in _minimize_replica
    logger.debug("Replica %5d/%5d: initial energy %8.3f kT", replica_index, self.nstates, state.reduced_potential(positions, box_vectors=box_vectors, context=context))
  File "/home/nadenl/miniconda3/envs/yex/lib/python3.5/site-packages/yank/repex.py", line 260, in reduced_potential
    potential_energy = self._compute_potential_energy(positions, box_vectors=box_vectors, platform=platform, context=context)
  File "/home/nadenl/miniconda3/envs/yex/lib/python3.5/site-packages/yank/repex.py", line 313, in _compute_potential_energy
    potential_energy = context.getState(getEnergy=True).getPotentialEnergy()
  File "/home/nadenl/miniconda3/envs/yex/lib/python3.5/site-packages/simtk/openmm/openmm.py", line 3969, in getState
    self._getStateAsLists(getP, getV, getF, getE, getPa, getPd, enforcePeriodic, groups_mask)
  File "/home/nadenl/miniconda3/envs/yex/lib/python3.5/site-packages/simtk/openmm/openmm.py", line 3919, in _getStateAsLists
    return _openmm.Context__getStateAsLists(self, getPositions, getVelocities, getForces, getEnergy, getParameters, getParameterDerivatives, enforcePeriodic, groups)
Exception: Error launching CUDA compiler: 256
<built-in>:0:0: fatal error: when writing output to : Bad file descriptor
compilation terminated.
```

**Comment by Lnaden** [comment]
Some additional testing notes:

This might actually be an MPI version problem. We are using `mpich2-1.4.1p1` (not sure which package installs it). However, `mpich2` got to version 1.5-ish then changed back in [November 2012](https://www.mpich.org/about/overview/) the versioning to just `mpich` starting at 3.0. There is now a 3.2 on conda-forge and it *appears* to not have this problem on the simple test. I will test with YANK itself first to see if that really is the problem.

**Comment by Lnaden** [comment]
Okay, I can confirm this is an MPI version problem. Mpich-3.2 fixes the problem

New problem:
* The `anaconda` (default) conda channel `mpi4py` installs `mpich2-1.41.p1` (the highest available version)
* We need `mpi4py` from the `conda-forge` channel to get `mpich-3.2` to fix the problem.
* There is no way to specify in the `meta.yaml` file to specify channels. 

Here are a few solutions to the problem:
1. Require people to install yank from `omnia`, then install mpi4py from `conda-forge`.
2. Require yank to be installed from `conda-forge` (probably want to wait until we have all of omnia).
3. Build our own `mpi4py` and `mpich` on `omina`

**Comment by Lnaden** [comment]
And we cannot script the first solution a part of the install since we don't know the conditions under which yank gets installed on a user-by-user basis.

We also cannot just require people to use an `mpi4py` version which would force the `conda-forge` version since the version on `anaconda` and `conda-forge` are the same

**Comment by Lnaden** [comment]
This should be fixed as #686, we can re-open if it crops up again
